[tools] AndroidAppBuilder dx dependence breaks on Android SDK 31.0.0 and up

```
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018: The "AndroidAppBuilderTask" task failed unexpectedly.
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018: System.ComponentModel.Win32Exception (2): An error occurred trying to start process '/Users/mdhwang/android-sdk/build-tools/32.0.0/dx' with working directory '/Users/mdhwang/runtime_droid/artifacts/tests/coreclr/obj/Android.x64.Debug/Managed/./build/AndroidApps/tracing_eventpipe/apk'. No such file or directory
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at System.Diagnostics.Process.Start()
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at Utils.TryRunProcess(TaskLoggingHelper logger, String path, String args, IDictionary`2 envVars, String workingDir, Boolean silent, Boolean logStdErrAsMessage, MessageImportance debugMessageImportance, String label) in /Users/mdhwang/runtime_droid/src/tasks/Common/Utils.cs:line 144
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at Utils.RunProcess(TaskLoggingHelper logger, String path, String args, IDictionary`2 envVars, String workingDir, Boolean ignoreErrors, Boolean silent, MessageImportance debugMessageImportance) in /Users/mdhwang/runtime_droid/src/tasks/Common/Utils.cs:line 88
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at ApkBuilder.BuildApk(String abi, String mainLibraryFileName, String monoRuntimeHeaders) in /Users/mdhwang/runtime_droid/src/tasks/AndroidAppBuilder/ApkBuilder.cs:line 387
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at AndroidAppBuilderTask.Execute() in /Users/mdhwang/runtime_droid/src/tasks/AndroidAppBuilder/AndroidAppBuilder.cs:line 110
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
/Users/mdhwang/runtime_droid/src/tests/build.proj(245,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

It looks like dx has officially been removed from Android SDK since version 31.0.0.
The plan to switch to d8 has finally finished its last phase https://android-developers.googleblog.com/2018/04/android-studio-switching-to-d8-dexer.html, so if we plan to support newer devices using these higher API levels, we should update the AndroidAppBuilder to work with d8.

--- a/src/tasks/AndroidAppBuilder/ApkBuilder.cs
+++ b/src/tasks/AndroidAppBuilder/ApkBuilder.cs
@@ -210,6 +210,7 @@ public class ApkBuilder
 
         // tools:
         string dx = Path.Combine(buildToolsFolder, "dx");
+        string d8 = Path.Combine(buildToolsFolder, "d8");
         string aapt = Path.Combine(buildToolsFolder, "aapt");
         string zipalign = Path.Combine(buildToolsFolder, "zipalign");
         string apksigner = Path.Combine(buildToolsFolder, "apksigner");
@@ -394,7 +395,20 @@ public class ApkBuilder
         string javaCompilerArgs = $"-d obj -classpath src -bootclasspath {androidJar} -source 1.8 -target 1.8 ";
         Utils.RunProcess(logger, javac, javaCompilerArgs + javaActivityPath, workingDir: OutputDir);
         Utils.RunProcess(logger, javac, javaCompilerArgs + monoRunnerPath, workingDir: OutputDir);
-        Utils.RunProcess(logger, dx, "--dex --output=classes.dex obj", workingDir: OutputDir);
+
+        if (File.Exists(d8))
+        {
+            string[] classFiles = Directory.GetFiles(Path.Combine(OutputDir, "obj"), "*.class", SearchOption.AllDirectories);
+
+            if (!classFiles.Any())
+                throw new InvalidOperationException("Didn't find any .class files");
+
+            Utils.RunProcess(logger, d8, $"--no-desugaring {string.Join(" ", classFiles)}", workingDir: OutputDir);
+        }
+        else
+        {
+            Utils.RunProcess(logger, dx, "--dex --output=classes.dex obj", workingDir: OutputDir);
+        }
 
         // 3. Generate APK
 

